src/async_callback.rs

```diff
@@ -0,0 +1,577 @@
+            Stack<'gc, '_>,
+        ) -> R,
+        R: 'seq,
+    {
+        visit_shared(move |shared| {
+            f(
+                shared.ctx,
+                Locals {
+                    locals: shared.locals,
+                    _invariant: PhantomData,
+                },
+                shared.exec.reborrow(),
+                shared.stack.reborrow(),
+            )
+        })
+    }
+
+    /// A version of `SequenceState::enter` which supports failure, and automatically turns any
+    /// returned error into a async sequence `Local`.
+    pub fn try_enter<F, R>(&mut self, f: F) -> Result<R, LocalError<'seq>>
+    where
+        F: for<'gc> FnOnce(
+            Context<'gc>,
+            Locals<'seq, 'gc>,
+            Execution<'gc, '_>,
+            Stack<'gc, '_>,
+        ) -> Result<R, Error<'gc>>,
+        R: 'seq,
+    {
+        visit_shared(move |shared| {
+            f(
+                shared.ctx,
+                Locals {
+                    locals: shared.locals,
+                    _invariant: PhantomData,
+                },
+                shared.exec.reborrow(),
+                shared.stack.reborrow(),
+            )
+            .map_err(|e| Local::stash(&shared.ctx, shared.locals, e))
+        })
+    }
+
+    /// Return `SequencePoll::Pending` to the code driving the `Sequence`.
+    ///
+    /// In normal use, this will return control to the calling `Executor` and potentially the
+    /// calling Rust code.
+    pub async fn pending(&mut self) {
+        wait_once().await;
+        visit_shared(move |shared| {
+            assert!(
+                shared.error.is_none(),
+                "SequencePoll::Pending cannot be followed by an error"
+            );
+        });
+    }
+
+    /// Finish the async sequence and return the values currently in the stack to the caller.
+    ///
+    /// Ending an async block *without* calling `SequenceState::return_to` will instead *drop* the
+    /// values in the stack.
+    pub fn return_to(self) {
+        visit_shared(move |shared| {
+            *shared.next = SequencePoll::Return;
+        });
+    }
+
+    /// Call the given Lua function with arguemnts / returns starting at `bottom` in the Stack.
+    pub async fn call(
+        &mut self,
+        func: &LocalFunction<'seq>,
+        bottom: usize,
+    ) -> Result<(), LocalError<'seq>> {
+        visit_shared(move |shared| {
+            *shared.next = SequencePoll::Call {
+                function: func.fetch(shared.locals),
+                bottom,
+            };
+        });
+        wait_once().await;
+        visit_shared(move |shared| {
+            if let Some(err) = shared.error.take() {
+                Err(Local::stash(&shared.ctx, shared.locals, err))
+            } else {
+                Ok(())
+            }
+        })
+    }
+
+    /// Finish the async sequence by calling the given Lua function with the arguments as the
+    /// current contents of the stack.
```

Comment:
Maybe "by calling the given Lua function with the current contents of the stack as arguments."?